File: python_cpp/uPIMulator_frontend/src/converter/op_code_converter.py
from abi.isa.instruction.op_code import OpCode


class OpCodeConverter:
    def __init__(self):
        pass

    @staticmethod
    def convert_to_op_code(op_code: str) -> OpCode:
        if op_code == "acquire":
            return OpCode.ACQUIRE
        elif op_code == "release":
            return OpCode.RELEASE
        elif op_code == "add":
            return OpCode.ADD
        elif op_code == "addc":
            return OpCode.ADDC
        elif op_code == "and":
            return OpCode.AND
        elif op_code == "andn":
            return OpCode.ANDN
        elif op_code == "asr":
            return OpCode.ASR
        elif op_code == "cao":
            return OpCode.CAO
        elif op_code == "clo":
            return OpCode.CLO
        elif op_code == "cls":
            return OpCode.CLS
        elif op_code == "clz":
            return OpCode.CLZ
        elif op_code == "cmpb4":
            return OpCode.CMPB4
        elif op_code == "div_step":
            return OpCode.DIV_STEP
        elif op_code == "extsb":
            return OpCode.EXTSB
        elif op_code == "extsh":
            return OpCode.EXTSH
        elif op_code == "extub":
            return OpCode.EXTUB
        elif op_code == "extuh":
            return OpCode.EXTUH
        elif op_code == "lsl":
            return OpCode.LSL
        elif op_code == "lsl_add":
            return OpCode.LSL_ADD
        elif op_code == "lsl_sub":
            return OpCode.LSL_SUB
        elif op_code == "lsl1":
            return OpCode.LSL1
        elif op_code == "lsl1x":
            return OpCode.LSL1X
        elif op_code == "lslx":
            return OpCode.LSLX
        elif op_code == "lsr":
            return OpCode.LSR
        elif op_code == "lsr_add":
            return OpCode.LSR_ADD
        elif op_code == "lsr1":
            return OpCode.LSR1
        elif op_code == "lsr1x":
            return OpCode.LSR1X
        elif op_code == "lsrx":
            return OpCode.LSRX
        elif op_code == "mul_sh_sh":
            return OpCode.MUL_SH_SH
        elif op_code == "mul_sh_sl":
            return OpCode.MUL_SH_SL
        elif op_code == "mul_sh_uh":
            return OpCode.MUL_SH_UH
        elif op_code == "mul_sh_ul":
            return OpCode.MUL_SH_UL
        elif op_code == "mul_sl_sh":
            return OpCode.MUL_SL_SH
        elif op_code == "mul_sl_sl":
            return OpCode.MUL_SL_SL
        elif op_code == "mul_sl_uh":
            return OpCode.MUL_SL_UH
        elif op_code == "mul_sl_ul":
            return OpCode.MUL_SL_UL
        elif op_code == "mul_step":
            return OpCode.MUL_STEP
        elif op_code == "mul_uh_uh":
            return OpCode.MUL_UH_UH
        elif op_code == "mul_uh_ul":
            return OpCode.MUL_UH_UL
        elif op_code == "mul_ul_uh":
            return OpCode.MUL_UL_UH
        elif op_code == "mul_ul_ul":
            return OpCode.MUL_UL_UL
        elif op_code == "nand":
            return OpCode.NAND
        elif op_code == "nor":
            return OpCode.NOR
        elif op_code == "nxor":
            return OpCode.NXOR
        elif op_code == "or":
            return OpCode.OR
        elif op_code == "orn":
            return OpCode.ORN
        elif op_code == "rol":
            return OpCode.ROL
        elif op_code == "rol_add":
            return OpCode.ROL_ADD
        elif op_code == "ror":
            return OpCode.ROR
        elif op_code == "rsub":
            return OpCode.RSUB
        elif op_code == "rsubc":
            return OpCode.RSUBC
        elif op_code == "sub":
            return OpCode.SUB
        elif op_code == "subc":
            return OpCode.SUBC
        elif op_code == "xor":
            return OpCode.XOR
        elif op_code == "boot":
            return OpCode.BOOT
        elif op_code == "resume":
            return OpCode.RESUME
        elif op_code == "stop":
            return OpCode.STOP
        elif op_code == "call":
            return OpCode.CALL
        elif op_code == "fault":
            return OpCode.FAULT
        elif op_code == "nop":
            return OpCode.NOP
        elif op_code == "sats":
            return OpCode.SATS
        elif op_code == "movd":
            return OpCode.MOVD
        elif op_code == "swapd":
            return OpCode.SWAPD
        elif op_code == "hash":
            return OpCode.HASH
        elif op_code == "time":
            return OpCode.TIME
        elif op_code == "time_cfg":
            return OpCode.TIME_CFG
        elif op_code == "lbs":
            return OpCode.LBS
        elif op_code == "lbu":
            return OpCode.LBU
        elif op_code == "ld":
            return OpCode.LD
        elif op_code == "lhs":
            return OpCode.LHS
        elif op_code == "lhu":
            return OpCode.LHU
        elif op_code == "lw":
            return OpCode.LW
        elif op_code == "sb":
            return OpCode.SB
        elif op_code == "sb_id":
            return OpCode.SB_ID
        elif op_code == "sd":
            return OpCode.SD
        elif op_code == "sd_id":
            return OpCode.SD_ID
        elif op_code == "sh":
            return OpCode.SH
        elif op_code == "sh_id":
            return OpCode.SH_ID
        elif op_code == "sw":
            return OpCode.SW
        elif op_code == "sw_id":
            return OpCode.SW_ID
        elif op_code == "ldma":
            return OpCode.LDMA
        elif op_code == "ldmai":
            return OpCode.LDMAI
        elif op_code == "sdma":
            return OpCode.SDMA
        else:
            raise ValueError

    @staticmethod
    def convert_to_string(op_code: OpCode) -> str:
        if op_code == OpCode.ACQUIRE:
            return "acquire"
        elif op_code == OpCode.RELEASE:
            return "release"
        elif op_code == OpCode.ADD:
            return "add"
        elif op_code == OpCode.ADDC:
            return "addc"
        elif op_code == OpCode.AND:
            return "and"
        elif op_code == OpCode.ANDN:
            return "andn"
        elif op_code == OpCode.ASR:
            return "asr"
        elif op_code == OpCode.CAO:
            return "cao"
        elif op_code == OpCode.CLO:
            return "clo"
        elif op_code == OpCode.CLS:
            return "cls"
        elif op_code == OpCode.CLZ:
            return "clz"
        elif op_code == OpCode.CMPB4:
            return "cmpb4"
        elif op_code == OpCode.DIV_STEP:
            return "div_step"
        elif op_code == OpCode.EXTSB:
            return "extsb"
        elif op_code == OpCode.EXTSH:
            return "extsh"
        elif op_code == OpCode.EXTUB:
            return "extub"
        elif op_code == OpCode.EXTUH:
            return "extuh"
        elif op_code == OpCode.LSL:
            return "lsl"
        elif op_code == OpCode.LSL_ADD:
            return "lsl_add"
        elif op_code == OpCode.LSL_SUB:
            return "lsl_sub"
        elif op_code == OpCode.LSL1:
            return "lsl1"
        elif op_code == OpCode.LSL1X:
            return "lsl1x"
        elif op_code == OpCode.LSLX:
            return "lslx"
        elif op_code == OpCode.LSR:
            return "lsr"
        elif op_code == OpCode.LSR_ADD:
            return "lsr_add"
        elif op_code == OpCode.LSR1:
            return "lsr1"
        elif op_code == OpCode.LSR1X:
            return "lsr1x"
        elif op_code == OpCode.LSRX:
            return "lsrx"
        elif op_code == OpCode.MUL_SH_SH:
            return "mul_sh_sh"
        elif op_code == OpCode.MUL_SH_SL:
            return "mul_sh_sl"
        elif op_code == OpCode.MUL_SH_UH:
            return "mul_sh_uh"
        elif op_code == OpCode.MUL_SH_UL:
            return "mul_sh_ul"
        elif op_code == OpCode.MUL_SL_SH:
            return "mul_sl_sh"
        elif op_code == OpCode.MUL_SL_SL:
            return "mul_sl_sl"
        elif op_code == OpCode.MUL_SL_UH:
            return "mul_sl_uh"
        elif op_code == OpCode.MUL_SL_UL:
            return "mul_sl_ul"
        elif op_code == OpCode.MUL_STEP:
            return "mul_step"
        elif op_code == OpCode.MUL_UH_UH:
            return "mul_uh_uh"
        elif op_code == OpCode.MUL_UH_UL:
            return "mul_uh_ul"
        elif op_code == OpCode.MUL_UL_UH:
            return "mul_ul_uh"
        elif op_code == OpCode.MUL_UL_UL:
            return "mul_ul_ul"
        elif op_code == OpCode.NAND:
            return "nand"
        elif op_code == OpCode.NOR:
            return "nor"
        elif op_code == OpCode.NXOR:
            return "nxor"
        elif op_code == OpCode.OR:
            return "or"
        elif op_code == OpCode.ORN:
            return "orn"
        elif op_code == OpCode.ROL:
            return "rol"
        elif op_code == OpCode.ROL_ADD:
            return "rol_add"
        elif op_code == OpCode.ROR:
            return "ror"
        elif op_code == OpCode.RSUB:
            return "rsub"
        elif op_code == OpCode.RSUBC:
            return "rsubc"
        elif op_code == OpCode.SUB:
            return "sub"
        elif op_code == OpCode.SUBC:
            return "subc"
        elif op_code == OpCode.XOR:
            return "xor"
        elif op_code == OpCode.BOOT:
            return "boot"
        elif op_code == OpCode.RESUME:
            return "resume"
        elif op_code == OpCode.STOP:
            return "stop"
        elif op_code == OpCode.CALL:
            return "call"
        elif op_code == OpCode.FAULT:
            return "fault"
        elif op_code == OpCode.NOP:
            return "nop"
        elif op_code == OpCode.SATS:
            return "sats"
        elif op_code == OpCode.MOVD:
            return "movd"
        elif op_code == OpCode.SWAPD:
            return "swapd"
        elif op_code == OpCode.HASH:
            return "hash"
        elif op_code == OpCode.TIME:
            return "time"
        elif op_code == OpCode.TIME_CFG:
            return "time_cfg"
        elif op_code == OpCode.LBS:
            return "lbs"
        elif op_code == OpCode.LBU:
            return "lbu"
        elif op_code == OpCode.LD:
            return "ld"
        elif op_code == OpCode.LHS:
            return "lhs"
        elif op_code == OpCode.LHU:
            return "lhu"
        elif op_code == OpCode.LW:
            return "lw"
        elif op_code == OpCode.SB:
            return "sb"
        elif op_code == OpCode.SB_ID:
            return "sb_id"
        elif op_code == OpCode.SD:
            return "sd"
        elif op_code == OpCode.SD_ID:
            return "sd_id"
        elif op_code == OpCode.SH:
            return "sh"
        elif op_code == OpCode.SH_ID:
            return "sh_id"
        elif op_code == OpCode.SW:
            return "sw"
        elif op_code == OpCode.SW_ID:
            return "sw_id"
        elif op_code == OpCode.LDMA:
            return "ldma"
        elif op_code == OpCode.LDMAI:
            return "ldmai"
        elif op_code == OpCode.SDMA:
            return "sdma"
        else:
            raise ValueError

File: tools/upmem_profiler/src/converter/op_code_converter.cc
#include "converter/op_code_converter.h"

#include <stdexcept>

namespace upmem_profiler::converter {

abi::instruction::OpCode OpCodeConverter::to_op_code(std::string op_code) {
  if (op_code == "acquire") {
    return abi::instruction::ACQUIRE;
  } else if (op_code == "release") {
    return abi::instruction::RELEASE;
  } else if (op_code == "add") {
    return abi::instruction::ADD;
  } else if (op_code == "addc") {
    return abi::instruction::ADDC;
  } else if (op_code == "and") {
    return abi::instruction::AND;
  } else if (op_code == "andn") {
    return abi::instruction::ANDN;
  } else if (op_code == "asr") {
    return abi::instruction::ASR;
  } else if (op_code == "cao") {
    return abi::instruction::CAO;
  } else if (op_code == "clo") {
    return abi::instruction::CLO;
  } else if (op_code == "cls") {
    return abi::instruction::CLS;
  } else if (op_code == "clz") {
    return abi::instruction::CLZ;
  } else if (op_code == "cmpb4") {
    return abi::instruction::CMPB4;
  } else if (op_code == "div_step") {
    return abi::instruction::DIV_STEP;
  } else if (op_code == "extsb") {
    return abi::instruction::EXTSB;
  } else if (op_code == "extsh") {
    return abi::instruction::EXTSH;
  } else if (op_code == "extub") {
    return abi::instruction::EXTUB;
  } else if (op_code == "extuh") {
    return abi::instruction::EXTUH;
  } else if (op_code == "lsl") {
    return abi::instruction::LSL;
  } else if (op_code == "lsl_add") {
    return abi::instruction::LSL_ADD;
  } else if (op_code == "lsl_sub") {
    return abi::instruction::LSL_SUB;
  } else if (op_code == "lsl1") {
    return abi::instruction::LSL1;
  } else if (op_code == "lsl1x") {
    return abi::instruction::LSL1X;
  } else if (op_code == "lslx") {
    return abi::instruction::LSLX;
  } else if (op_code == "lsr") {
    return abi::instruction::LSR;
  } else if (op_code == "lsr_add") {
    return abi::instruction::LSR_ADD;
  } else if (op_code == "lsr1") {
    return abi::instruction::LSR1;
  } else if (op_code == "lsr1x") {
    return abi::instruction::LSR1X;
  } else if (op_code == "lsrx") {
    return abi::instruction::LSRX;
  } else if (op_code == "mul_sh_sh") {
    return abi::instruction::MUL_SH_SH;
  } else if (op_code == "mul_sh_sl") {
    return abi::instruction::MUL_SH_SL;
  } else if (op_code == "mul_sh_uh") {
    return abi::instruction::MUL_SH_UH;
  } else if (op_code == "mul_sh_ul") {
    return abi::instruction::MUL_SH_UL;
  } else if (op_code == "mul_sl_sh") {
    return abi::instruction::MUL_SL_SH;
  } else if (op_code == "mul_sl_sl") {
    return abi::instruction::MUL_SL_SL;
  } else if (op_code == "mul_sl_uh") {
    return abi::instruction::MUL_SL_UH;
  } else if (op_code == "mul_sl_ul") {
    return abi::instruction::MUL_SL_UL;
  } else if (op_code == "mul_step") {
    return abi::instruction::MUL_STEP;
  } else if (op_code == "mul_uh_uh") {
    return abi::instruction::MUL_UH_UH;
  } else if (op_code == "mul_uh_ul") {
    return abi::instruction::MUL_UH_UL;
  } else if (op_code == "mul_ul_uh") {
    return abi::instruction::MUL_UL_UH;
  } else if (op_code == "mul_ul_ul") {
    return abi::instruction::MUL_UL_UL;
  } else if (op_code == "nand") {
    return abi::instruction::NAND;
  } else if (op_code == "nor") {
    return abi::instruction::NOR;
  } else if (op_code == "nxor") {
    return abi::instruction::NXOR;
  } else if (op_code == "or") {
    return abi::instruction::OR;
  } else if (op_code == "orn") {
    return abi::instruction::ORN;
  } else if (op_code == "rol") {
    return abi::instruction::ROL;
  } else if (op_code == "rol_add") {
    return abi::instruction::ROL_ADD;
  } else if (op_code == "ror") {
    return abi::instruction::ROR;
  } else if (op_code == "rsub") {
    return abi::instruction::RSUB;
  } else if (op_code == "rsubc") {
    return abi::instruction::RSUBC;
  } else if (op_code == "sub") {
    return abi::instruction::SUB;
  } else if (op_code == "subc") {
    return abi::instruction::SUBC;
  } else if (op_code == "xor") {
    return abi::instruction::XOR;
  } else if (op_code == "boot") {
    return abi::instruction::BOOT;
  } else if (op_code == "resume") {
    return abi::instruction::RESUME;
  } else if (op_code == "stop") {
    return abi::instruction::STOP;
  } else if (op_code == "call") {
    return abi::instruction::CALL;
  } else if (op_code == "fault") {
    return abi::instruction::FAULT;
  } else if (op_code == "nop") {
    return abi::instruction::NOP;
  } else if (op_code == "sats") {
    return abi::instruction::SATS;
  } else if (op_code == "movd") {
    return abi::instruction::MOVD;
  } else if (op_code == "swapd") {
    return abi::instruction::SWAPD;
  } else if (op_code == "hash") {
    return abi::instruction::HASH;
  } else if (op_code == "time") {
    return abi::instruction::TIME;
  } else if (op_code == "time_cfg") {
    return abi::instruction::TIME_CFG;
  } else if (op_code == "lbs") {
    return abi::instruction::LBS;
  } else if (op_code == "lbu") {
    return abi::instruction::LBU;
  } else if (op_code == "ld") {
    return abi::instruction::LD;
  } else if (op_code == "lhs") {
    return abi::instruction::LHS;
  } else if (op_code == "lhu") {
    return abi::instruction::LHU;
  } else if (op_code == "lw") {
    return abi::instruction::LW;
  } else if (op_code == "sb") {
    return abi::instruction::SB;
  } else if (op_code == "sb_id") {
    return abi::instruction::SB_ID;
  } else if (op_code == "sd") {
    return abi::instruction::SD;
  } else if (op_code == "sd_id") {
    return abi::instruction::SD_ID;
  } else if (op_code == "sh") {
    return abi::instruction::SH;
  } else if (op_code == "SH_ID") {
    return abi::instruction::SH_ID;
  } else if (op_code == "sw") {
    return abi::instruction::SW;
  } else if (op_code == "sw_id") {
    return abi::instruction::SW_ID;
  } else if (op_code == "ldma") {
    return abi::instruction::LDMA;
  } else if (op_code == "ldmai") {
    return abi::instruction::LDMAI;
  } else if (op_code == "sdma") {
    return abi::instruction::SDMA;
  } else {
    throw std::invalid_argument("");
  }
}

} // namespace upmem_profiler::converter
